language/vm/vm-runtime/src/libra_vm.rs

```diff
@@ -557,15 +556,23 @@ pub fn chunk_block_transactions(txns: Vec<Transaction>) -> Vec<TransactionBlock>
                 }
                 blocks.push(TransactionBlock::BlockPrologue(data));
             }
-            Transaction::WriteSet(ws) => {
+            Transaction::WriteSet(cs) => {
                 if !buf.is_empty() {
                     blocks.push(TransactionBlock::UserTransaction(buf));
                     buf = vec![];
                 }
-                blocks.push(TransactionBlock::WriteSet(ws));
+                blocks.push(TransactionBlock::WriteSet(cs));
             }
             Transaction::UserTransaction(txn) => {
-                buf.push(txn);
+                if let TransactionPayload::WriteSet(cs) = txn.payload() {
+                    if !buf.is_empty() {
+                        blocks.push(TransactionBlock::UserTransaction(buf));
+                        buf = vec![];
+                    }
+                    blocks.push(TransactionBlock::WriteSet(cs.clone()));
```

Comment:
just a quick comment here: we should be able to remove this clone right? I mean we own the transaction here...

Reply:
The problem is that `payload` only returns a reference. And we couldn't call the `into_payload` here due to the else branch. It's a really nasty case but I don't see an easy way of getting around that...